qiskit/transpiler/synthesis/aqc/aqc.py

```diff
@@ -0,0 +1,118 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Main entry point to Approximate Quantum Compiler.
+"""
+from typing import Optional
+
+import numpy as np
+
+from .optimizers import GDOptimizer
+from .parametric_circuit import ParametricCircuit
+
+
+class AQC:
+    r"""
+    Implementation of Approximate Quantum Compiler as described in the paper.
+
+    We are interested in compiling a quantum circuit, which we formalize as finding the best
+    circuit representation in terms of an ordered gate sequence of a target unitary matrix
+    :math:`U\in \U(d)`, with some additional hardware constraints. In particular, we look at
+    representations that could be constrained in terms of hardware connectivity, as well
+    as gate depth, and we choose a gate basis in terms of CNOT and rotation gates.
+    We recall that the combination of CNOT and rotation gates is universal in :math:`\SU(d)` and
+    therefore it does not limit compilation.
+
+    To properly define what we mean by best circuit representation, we define the metric
+    as the Frobenius norm between the unitary matrix of the compiled circuit :math:`V` and
+    the target unitary matrix :math:`U`, i.e., :math:`\|V - U\|_{\mathrm{F}}`. This choice
+    is motivated by mathematical programming considerations, and it is related to other
+    formulations that appear in the literature.
+
+    References:
+
+        [1]: Liam Madden, Andrea Simonetto, Best Approximate Quantum Compiling Problems.
+            `arXiv:2106.05649 <https://arxiv.org/abs/2106.05649>`_
+    """
+
+    def __init__(
+        self,
+        method: str = "nesterov",
+        maxiter: int = 100,
+        eta: float = 0.1,
+        tol: float = 1e-5,
+        eps: float = 0,
```

Comment:
If I understand correctly, the process is stochastic and, the same amount of iterations, might result in significant different runtimes. In similar passes, qiskit allows to set a timeout. Even if the iteration was not reached, if the timeout is triggered, the best solution so far can be returned.

```suggestion
        eps: float = 0,
        timeout = None,
```

Reply:
Now, a custom made gradient descent is replaced with `qiskit.algorithms.optimizers.Optimizer` and the sub-classes. You pass an instance of an optimizer and that's it. You can tune it in a way you like it. If the optimizer supports a timeout parameter, then it can be passed to the optimizer. If not, then the only option is `max_iter` parameter to manage optimization time.